### PR TITLE
Add Micropayment channel

### PR DESCRIPTION
**Pull request by amitkumarj441**
Signed-off-by: Amit Kumar Jaiswal <amitkumarj441@gmail.com>

### CODE REVIEW

**Comment by axic** [comment]
Thanks! Can you please compare this with #2431 which is another sample code for the same?

**Comment by amitkumarj441** [comment]
@axic Thanks for pointing it out. But that's a sample code and here is the one I created one for `Payment restrict to only sender or recipient`.

Please review this PR.

**Comment by axic** [comment]
Sorry @amitkumarj441, I am not sure what do you mean by `Payment restrict to only sender or recipient` ?

**Comment by amitkumarj441** [comment]
@axic I mean it's for Payment Channel.

File: docs/solidity-by-example.rst [hunk 1]
```diff
@@ -618,4 +618,66 @@ Safe Remote Purchase
...
+        // `r`, `s`, and `v` (additional 2 bits). Arguments include:
+        // - `h` "data" Hash
+        // - `r` "signature" Hex values (0-65) of signature (i.e. `0x..`)
+        // - `s` "signature" Hex values (66-129) of signature (i.e. `0x` + `..`)
+        // - `v` "public key" Hex values (130-131) of signature (i.e. `0x` + `..`)
+        // Obtain signer from signature for verification that matches only sender or recipient
+        signer = ecrecover(h, v, r, s);
+
+        // Signature is invalid so throw
+        if (signer != channelSender && signer != channelRecipient) throw;
```
Comment (VoR0220) [inline docs/solidity-by-example.rst:660]: Needs to use revert rather than throw. Or use an assertion.

File: docs/solidity-by-example.rst [hunk 2]
```diff
@@ -618,4 +618,66 @@ Safe Remote Purchase
...
+
+        proof = sha3(this, value);
+
+        // Signature valid but not match data provided
+        if (proof != h) throw;
```
Comment (VoR0220) [inline docs/solidity-by-example.rst:665]: See above

File: docs/solidity-by-example.rst [hunk 3]
```diff
@@ -618,4 +618,66 @@ Safe Remote Purchase
...
+
+        if (signatures[proof] == 0)
+            signatures[proof] = signer;
+        else if (signatures[proof] != signer){
+            // Channel completed and both signatures provided
+            if (!channelRecipient.send(value)) throw;
```
Comment (VoR0220) [inline docs/solidity-by-example.rst:671]: And this definitely needs better construction right here. Perhaps using a modifier.

File: docs/solidity-by-example.rst [hunk 4]
```diff
@@ -618,4 +618,66 @@ Safe Remote Purchase
...
+            selfdestruct(channelSender);
+        }
+
+    }
+
+    function ChannelTimeout(){
+        if (startDate + channelTimeout > now)
```
Comment (VoR0220) [inline docs/solidity-by-example.rst]: this should be ```if (startDate + channelTimeout > now) selfdestruct(channelSender);```

Comment (amitkumarj441) [reply docs/solidity-by-example.rst]: Alright @VoR0220 . I just updated it.
Thanks!

**Comment by axic** [comment]
@amitkumarj441 sorry I am still confused about differences between this and #2431?

**Comment by amitkumarj441** [comment]
@axic This PR is an improved version of #2431. @VoR0220 is already reviewing this PR.

**Comment by axic** [comment]
To me this seems to be a code from scratch, not an incremental improvement over #2431.

I think having the state enum in #2431 is very useful. This PR also should adhere strictly to the coding guide as seen in https://solidity.readthedocs.io/en/develop/style-guide.html

**Comment by VoR0220** [comment]
This is a fairly basic one, I do think there are things that need to be altered in terms of a modifier (an enum works too frankly for state transition steps), however I did find it to be a tad more complex in terms of the details. Thorough would be the correct word but harder to decipher.

**Comment by amitkumarj441** [comment]
@VoR0220 Okay. Alright.
So, any decision on this PR?

**Comment by VoR0220** [comment]
Not my decision to make.

File: docs/solidity-by-example.rst (hunk 3 above)
Comment (axic) [reply docs/solidity-by-example.rst:671]: `.transfer()` is the safe preferred alternative.

**Comment by axic** [comment]
I am sorry, meanwhile #4212 was merged which achieves the same goal. Closing this now, but if you feel that the current merged chapter could be improved, please do not hesitate to open a PR for it.

**Pull request closed without merging**